### PR TITLE
Merging the HIP and CUDA unit tests into one

### PR DESCRIPTION
**Pull request by deven-amd**
Note:

1. `cxx11_tensor_*_[hip|cuda].cu` are renamed to `cxx11_tensor_*_gpu.cu`
2. The `cuda_basic` and `hip_basic` unit test are being kept separate for the time being.

Please review the `if !defined(EIGEN_USE_HIP)` parts ... basically they identify the subtests that either fail/hang on HIP and hence need fixing (there are just a few of them)

This is the last major change we should need to do!

### CODE REVIEW

**Review by whchung — COMMENTED** [review]
It seems some of the test are renamed from `_cuda.cu` and some of them are from `_hip.cu`. What is the rationale behind them?

File: cmake/EigenTesting.cmake [hunk 1]
```diff
@@ -20,7 +20,8 @@ macro(ei_add_test_internal testname testname_with_suffix)
 
   if(EIGEN_ADD_TEST_FILENAME_EXTENSION STREQUAL cu)
     if(EIGEN_TEST_HIP)
-      hip_add_executable(${targetname} ${filename} HIPCC_OPTIONS "-DEIGEN_USE_HIP")
+      hip_reset_flags()
```
Comment (whchung) [inline cmake/EigenTesting.cmake:23]: what is the purpose of `hip_reset_flags()` here?

**Review by whchung — COMMENTED** [review]
Also I believe we should be checking places where `EIGEN_USE_HIP` instead of `EIGEN_HIPCC`?

File: unsupported/test/CMakeLists.txt [hunk 2]
```diff
@@ -319,17 +319,18 @@ if (EIGEN_TEST_HIP)
...
-	ei_add_test(cxx11_tensor_reduction_hip)
-	ei_add_test(cxx11_tensor_argmax_hip)
-	ei_add_test(cxx11_tensor_cast_float16_hip)
-	ei_add_test(cxx11_tensor_scan_hip)
-	ei_add_test(cxx11_tensor_device_hip)
-	ei_add_test(cxx11_tensor_hip)
-	ei_add_test(cxx11_tensor_contract_hip)
-	ei_add_test(cxx11_tensor_of_float16_hip)
-	ei_add_test(cxx11_tensor_random_hip)
+	# ei_add_test(cxx11_tensor_complex_gpu)
```
Comment (whchung) [inline unsupported/test/CMakeLists.txt]: we probably want to add comment that these tests aren't supported in HIP yet.

**Review by whchung — COMMENTED** [review]
@deven-amd great job! in general this PR looks good to me. just some minor editorial changes.

File: cmake/EigenTesting.cmake (hunk 1 above)
Comment (deven-amd) [reply cmake/EigenTesting.cmake:23]: that was the key to making the `CALL_SUBTEST_{N}` macros work correctly.

`$ARGV2` contains `-DEIGEN_TEST_PART_{N}`, which in turn will enable the corresponding `CALL_SUBTEST_{N}` (and disable the rest). 

Without the call to `hip_reset_flags` , HIPCC_OPTIONS was somehow getting corrupted and the correct `-DEIGEN_TEST_PART_{N}` was not getting specified on the generated command line. This is why I had lumped all subtests into one (i.e. removed the `_{N}` from the `CALL_SUBTEST_{N}`) in the `_hip` versions of the tests. After some digging around in `FindHIP.cmake`, I tried `hip_reset_flags` on a hunch and it seems to have gotten things working.

File: unsupported/test/CMakeLists.txt (hunk 2 above)
Comment (deven-amd) [reply unsupported/test/CMakeLists.txt]: yup...will add those comments in and update PR later tonight / tomorrow

**Comment by deven-amd** [comment]
regarding the renaming to `_gpu.cu`.  It was only `_cuda.cu` files that got renamed to `_gpu.cu` and the `_hip.cu` ones were the ones that were deleted. However it seems that if you do the rename and delete in the same commit, git gets confused (possibly because the `_hip.cu` versions were initially copied from the `_cuda.cu` versions). I had seen the same thing happen when I renamed the `Tensor*_cuda.h` files. I had to break up the renaming and deleting into separate commits to get around it. Forgot to do that this time around!

**Comment by deven-amd** [comment]
I do not think I quite follow you on the `EIGEN_USE_HIP` vs `EIGEN_HIPCC` issue...can you elaborate? thanks

File: cmake/EigenTesting.cmake (hunk 1 above)
Comment (whchung) [reply cmake/EigenTesting.cmake:23]: thanks for the clarification

**Comment by whchung** [comment]
@deven-amd in the description of the PR you mentioned to look for `#if defined(EIGEN_HIPCC)` but i only saw `#if !defined(EIGEN_USE_HIP)`?

**Comment by deven-amd** [comment]
@whchung 

I see what you were referring to...you are right. The PR description should have read `#if !defined(EIGEN_USE_HIP` instead of `#if defined(EIGEN_HIPCC)`.  I am updating the description to reflect that.

Also I updated the PR with the comment added for skipping the complex datatype unit tests for HIP.

Please review and proceed with merge.

Thanks again

deven

**Pull request merged**